git.go

```diff
@@ -0,0 +1,244 @@
+}
+
+func run(c *command.Model) error {
+	log.Infof(c.PrintableCommandArgs())
+	return c.SetStdout(os.Stdout).SetStderr(os.Stderr).Run()
+}
+
+func runForOutput(c *command.Model) (string, error) {
+	// log.Infof(c.PrintableCommandArgs())
+	return c.RunAndReturnTrimmedCombinedOutput()
+}
+
+func runWithRetry(f func() *command.Model) error {
+	return retry.Times(retryCount).Wait(waitTime).Try(func(attempt uint) error {
+		if attempt > 0 {
+			log.Warnf("Retrying...")
+		}
+
+		err := run(f())
+		if err != nil {
+			log.Warnf("Attempt %d failed:", attempt+1)
+			fmt.Println(err.Error())
+		}
+
+		return err
+	})
+}
+
+func isFork() bool {
+	return config.PRRepositoryCloneURL != "" &&
+		config.RepositoryURL != config.PRRepositoryCloneURL
+}
+
+func isPrivate() bool {
+	return strings.HasPrefix(config.PRRepositoryCloneURL, "git")
+}
+
+func autoMerge() error {
+	if err := runWithRetry(func() *command.Model {
+		if config.CloneDepth != "" {
+			return Git.Fetch("--depth=" + config.CloneDepth)
+		}
+		return Git.Fetch()
+	}); err != nil {
+		return fmt.Errorf("Fetch failed, error: %v", err)
+	}
+
+	if config.PRMergeBranch != "" {
+		if err := runWithRetry(func() *command.Model {
+			return Git.Fetch("origin", config.PRMergeBranch+":"+
+				strings.TrimSuffix(config.PRMergeBranch, "/merge"))
+		}); err != nil {
+			return fmt.Errorf("fetch Pull Request branch failed (%s), error: %v",
+				config.PRMergeBranch, err)
+		}
+
+		arg := strings.TrimSuffix(config.PRMergeBranch, "/merge")
+		if err := run(Git.Checkout(arg)); err != nil {
+			return fmt.Errorf("checkout failed (%s), error: %v", config.BranchDest, err)
+		}
+	} else if patch, err := getDiffFile(); err == nil {
+		if err := run(Git.Checkout(config.BranchDest)); err != nil {
+			return fmt.Errorf("checkout failed (%s), error: %v", config.BranchDest, err)
+		}
+		if err := run(Git.Apply(patch)); err != nil {
+			return fmt.Errorf("can't apply patch (%s), error: %v", patch, err)
+		}
+	} else {
+		return fmt.Errorf("there is no Pull Request branch and can't download diff file")
+	}
+	return nil
+}
+
+func manualMerge() error {
+	if err := runWithRetry(func() *command.Model {
+		if config.CloneDepth != "" {
+			return Git.Fetch("--depth=" + config.CloneDepth)
+		}
+		return Git.Fetch()
+	}); err != nil {
+		return fmt.Errorf("Fetch failed, error: %v", err)
+	}
+
+	if err := run(Git.Checkout(config.BranchDest)); err != nil {
+		return fmt.Errorf("checkout failed (%s), error: %v", config.BranchDest, err)
+	}
+
+	if isFork() {
+		if err := run(Git.RemoteAdd("upstream", config.PRRepositoryCloneURL)); err != nil {
+			return fmt.Errorf("couldn't add remote (%s), error: %v", config.PRRepositoryCloneURL, err)
+		}
+
+		if err := runWithRetry(func() *command.Model {
+			return Git.Fetch("upstream", config.Branch)
+		}); err != nil {
+			return fmt.Errorf("fetch Pull Request branch failed (%s), error: %v",
+				config.Branch, err)
+		}
+
+		if err := run(Git.Merge("upstream/" + config.Branch)); err != nil {
```

Comment:
an here of course - and everywhere where "upstream" is used, as we changed the order